Source/Core/Core/AchievementManager.cpp

```diff
@@ -146,80 +154,210 @@ void AchievementManager::LoadGameByFilenameAsync(const std::string& iso_path,
+              // Unused
+              break;
+            }
+          },
+      .tell =
+          [](void* file_handle) {
+            return reinterpret_cast<FilereaderState*>(file_handle)->position;
+          },
+      .read =
+          [](void* file_handle, void* buffer, size_t requested_bytes) {
+            FilereaderState* filereader_state = reinterpret_cast<FilereaderState*>(file_handle);
+            bool success = (filereader_state->volume->Read(
+                filereader_state->position, requested_bytes, reinterpret_cast<u8*>(buffer),
+                DiscIO::PARTITION_NONE));
+            if (success)
+            {
+              filereader_state->position += requested_bytes;
+              return requested_bytes;
+            }
+            else
+            {
+              return static_cast<size_t>(0);
+            }
+          },
+      .close = [](void* file_handle) { delete reinterpret_cast<FilereaderState*>(file_handle); }};
+  rc_hash_init_custom_filereader(&volume_reader);
   m_queue.EmplaceItem([this, callback] {
-    const auto resolve_hash_response = ResolveHash(this->m_game_hash);
-    if (resolve_hash_response != ResponseType::SUCCESS || m_game_id == 0)
+    Hash new_hash;
+    if (!rc_hash_generate_from_file(new_hash.data(), RC_CONSOLE_GAMECUBE, ""))
     {
-      callback(resolve_hash_response);
-      INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
-      OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
+      ERROR_LOG_FMT(ACHIEVEMENTS, "Unable to generate achievement hash from volume.");
+      callback(AchievementManager::ResponseType::MALFORMED_OBJECT);
       return;
     }
-
-    const auto start_session_response = StartRASession();
-    if (start_session_response != ResponseType::SUCCESS)
     {
-      callback(start_session_response);
-      WARN_LOG_FMT(ACHIEVEMENTS, "Failed to connect to RetroAchievements server.");
-      OSD::AddMessage("Failed to connect to RetroAchievements server.", OSD::Duration::VERY_LONG,
-                      OSD::Color::RED);
-      return;
+      std::lock_guard lg{m_lock};
+      m_game_hash = std::move(new_hash);
     }
+    delete m_loading_volume;
+    m_loading_volume = nullptr;
+    LoadGameSync(callback);
+  });
+}
+
+void AchievementManager::LoadGameSync(const ResponseCallback& callback)
+{
+  u32 new_game_id = 0;
+  Hash current_hash;
+  {
+    std::lock_guard lg{m_lock};
+    current_hash = m_game_hash;
+  }
+  const auto resolve_hash_response = ResolveHash(this->m_game_hash, &new_game_id);
+  if (resolve_hash_response != ResponseType::SUCCESS || new_game_id == 0)
+  {
+    callback(resolve_hash_response);
+    INFO_LOG_FMT(ACHIEVEMENTS, "No RetroAchievements data found for this game.");
+    OSD::AddMessage("No RetroAchievements data found for this game.", OSD::Duration::VERY_LONG,
+                    OSD::Color::RED);
+    return;
```

Comment:
On the one hand, if the user switches from a known game to an unknown game this will get picked up the same way as switching from a known game to a different known game. On the other, once the system has booted and there's not achievement data found, it's not like it will ever reach the uncompromised state until the system reboots, so I could definitely set the flag here. But I might need to rethink some of my messaging if I make this change.

Reply:
> On the one hand, if the user switches from a known game to an unknown game this will get picked up the same way as switching from a known game to a different known game.

As far as I can tel, in the former case you don't set the compromised flag but in the latter case you do. Or am I missing something?

Reply:
I spoke too soon; you're right.